src/CurveCallee.sol

```
@@ -0,0 +1,182 @@
+        bytes calldata data        // Extra data, see below
+    ) external {
+        (
+            address to,            // address to send remaining DAI to
+            address gemJoin,       // gemJoin adapter address
+            uint256 minProfit,     // minimum profit in DAI to make [wad]
+             uint24 poolFee,       // uniswap V3 WETH-DAI pool fee
+            address charterManager // pass address(0) if no manager
+        ) = abi.decode(data, (address, address, uint256, uint24, address));
+
+        address gem = GemJoinLike(gemJoin).gem();
+        require(keccak256(bytes(TokenLike(gem).symbol())) == keccak256("wstETH"), "CurveCallee: only-wsteth");
+
+        // Convert slice to token precision
+        slice = _fromWad(gemJoin, slice);
+
+        // Exit gem to token
+        if(charterManager != address(0)) {
+            CharterManagerLike(charterManager).exit(gemJoin, address(this), slice);
+        } else {
+            GemJoinLike(gemJoin).exit(address(this), slice);
+        }
+
+        slice = WstEthLike(gem).unwrap(slice);
+        gem = WstEthLike(gem).stETH();
+
+        TokenLike(gem).approve(address(curve), slice);
+        slice = curve.exchange({
+                 i: 1,     // send token id 1 (stETH)
+                 j: 0,     // receive token id 0 (ETH)
+                dx: slice, // send `slice` amount of stETH
+            min_dy: 0      // accept any amount of ETH (`minProfit` is checked below)
+        });
+
+        gem = uniV3.WETH9();
+        WethLike(gem).deposit{
+            value: slice
+        }();
+
+        // Approve uniV3 to take gem
+        WethLike(gem).approve(address(uniV3), slice);
+
+        // Calculate amount of DAI to Join (as erc20 WAD value)
+        uint256 daiToJoin = divup(owe, RAY);
+
+        // Do operation and get dai amount bought (checking the profit is achieved)
+        bytes memory path = abi.encodePacked(gem, poolFee, address(dai));
```

Why not get the uniswap path as input instead of encoding it on chain? Similar to the uniswap callees.

---

Because in this case it will always be the same, so no need to get it as a parameter.

---

You are right, `poolFee` may change as there are at least two ETH/DAI pools on UniV3. That is why it is a parameter. I still don’t get why you want to avoid doing the encoding on-chain.

---

To save the gas of course.
We already get a Univ3 path in UniV3Calle.sol, so no reason for it to be different.

---

This is a memory opération. The amount of gas saved is probably négligible. It does make sense for it to be différent, since this is a contract spécifique for wstETH. Asking the user to provide a path, but silently failing if they don’t provide a spécifique path seems counterintuitive to me.